instrumentation/spring/spring-integration-4.1/testing/src/main/java/io/opentelemetry/javaagent/instrumentation/spring/integration/v4_1/RabbitExtension.java

```diff
@@ -0,0 +1,137 @@
+import org.springframework.cloud.stream.messaging.Source;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.context.annotation.Bean;
+import org.springframework.messaging.support.MessageBuilder;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.wait.strategy.Wait;
+
+public class RabbitExtension implements BeforeEachCallback, AfterEachCallback {
+
+  private GenericContainer<?> rabbitMqContainer;
+  protected ConfigurableApplicationContext producerContext;
+  private ConfigurableApplicationContext consumerContext;
+
+  private final InstrumentationExtension testing;
+  private final Class<?> additionalContextClass;
+
+  public RabbitExtension(InstrumentationExtension testing, Class<?> additionalContextClass) {
+    this.testing = testing;
+    this.additionalContextClass = additionalContextClass;
+  }
+
+  public <T> T getBean(String name, Class<T> requiredType) {
+    return producerContext.getBean(name, requiredType);
+  }
+
+  @Override
+  public void beforeEach(ExtensionContext context) {
+    rabbitMqContainer =
+        new GenericContainer<>("rabbitmq:latest")
+            .withExposedPorts(5672)
+            .waitingFor(Wait.forLogMessage(".*Server startup complete.*", 1))
+            .withStartupTimeout(Duration.ofMinutes(2));
+    rabbitMqContainer.start();
+
+    SpringApplication producerApp =
+        new SpringApplication(getContextClasses(ProducerConfig.class, additionalContextClass));
+    Map<String, Object> producerProperties = new HashMap<>();
+    producerProperties.put("spring.application.name", "testProducer");
+    producerProperties.put("spring.jmx.enabled", false);
+    producerProperties.put("spring.main.web-application-type", "none");
+    producerProperties.put("spring.rabbitmq.host", rabbitMqContainer.getHost());
+    producerProperties.put("spring.rabbitmq.port", rabbitMqContainer.getMappedPort(5672));
+    producerProperties.put("spring.cloud.stream.bindings.output.destination", "testTopic");
+    producerApp.setDefaultProperties(producerProperties);
+    producerContext = producerApp.run();
+
+    SpringApplication consumerApp =
+        new SpringApplication(
+            getContextClasses(ProducerConfig.ConsumerConfig.class, additionalContextClass));
+    Map<String, Object> consumerProperties = new HashMap<>();
+    consumerProperties.put("spring.application.name", "testConsumer");
+    consumerProperties.put("spring.jmx.enabled", false);
+    consumerProperties.put("spring.main.web-application-type", "none");
+    consumerProperties.put("spring.rabbitmq.host", rabbitMqContainer.getHost());
+    consumerProperties.put("spring.rabbitmq.port", rabbitMqContainer.getMappedPort(5672));
+    consumerProperties.put("spring.cloud.stream.bindings.input.destination", "testTopic");
+    consumerApp.setDefaultProperties(consumerProperties);
+    consumerContext = consumerApp.run();
+
+    testing.clearData();
```

Comment:
it might not be needed, but in other tests we have used `testing.waitForTraces(n)` to explicitly wait for the number of expected spans before clearing and continuing to prevent flakyness from any weird race conditions

Reply:
good idea